internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,274 @@
+	autoScalerEventPollerSkipUpdateNoChange                   = "skip-update-poller-limit-no-change"
+	autoScalerEventPollerSkipUpdateNotEnabled                 = "skip-update-poller-limit-not-enabled"
+	autoScalerEventMetrics                                    = "metrics"
+	autoScalerEventEnable                                     = "enable"
+	autoScalerEventDisable                                    = "disable"
+	autoScalerEventStart                                      = "start"
+	autoScalerEventStop                                       = "stop"
+	autoScalerEventLogMsg                     string          = "concurrency auto scaler event"
+	testTimeFormat                            string          = "15:04:05"
+
+	metricsEnabled        = "enabled"
+	metricsDisabled       = "disabled"
+	metricsPollerQuota    = "poller-quota"
+	metricsPollerWaitTime = "poller-wait-time"
+)
+
+type (
+	ConcurrencyAutoScaler struct {
+		shutdownChan chan struct{}
+		wg           sync.WaitGroup
+		log          *zap.Logger
+		scope        tally.Scope
+		clock        clockwork.Clock
+
+		concurrency *ConcurrencyLimit
+		cooldown    time.Duration
+		updateTick  time.Duration
+
+		// state of autoscaler
+		lock    sync.RWMutex
+		enabled bool
+
+		// poller
+		pollerInitCount        int
+		pollerMaxCount         int
+		pollerMinCount         int
+		pollerWaitTime         *rollingAverage[time.Duration]
+		pollerPermitLastUpdate time.Time
+	}
+
+	ConcurrencyAutoScalerInput struct {
+		Concurrency    *ConcurrencyLimit
+		Cooldown       time.Duration // cooldown time of update
+		Tick           time.Duration // frequency of update check
+		PollerMaxCount int
+		PollerMinCount int
+		Logger         *zap.Logger
+		Scope          tally.Scope
+		Clock          clockwork.Clock
+	}
+
+	autoScalerEvent string
+)
+
+func NewConcurrencyAutoScaler(input ConcurrencyAutoScalerInput) *ConcurrencyAutoScaler {
+	tick := defaultAutoScalerUpdateTick
+	if input.Tick != 0 {
+		tick = input.Tick
+	}
+	return &ConcurrencyAutoScaler{
+		shutdownChan:           make(chan struct{}),
+		concurrency:            input.Concurrency,
+		cooldown:               input.Cooldown,
+		log:                    input.Logger.Named(metrics.ConcurrencyAutoScalerScope),
+		scope:                  input.Scope.SubScope(metrics.ConcurrencyAutoScalerScope),
+		clock:                  input.Clock,
+		updateTick:             tick,
+		enabled:                false, // initial value should be false and is only turned on from auto config hint
+		pollerInitCount:        input.Concurrency.PollerPermit.Quota(),
+		pollerMaxCount:         input.PollerMaxCount,
+		pollerMinCount:         input.PollerMinCount,
+		pollerWaitTime:         newRollingAverage[time.Duration](numberOfPollsInRollingAverage),
+		pollerPermitLastUpdate: input.Clock.Now(),
+	}
+}
+
+func (c *ConcurrencyAutoScaler) Start() {
+	c.logEvent(autoScalerEventStart)
+
+	c.wg.Add(1)
+
+	go func() {
+		defer c.wg.Done()
```

Comment:
nit: any calls to go should have a panic handler.
If a bug exists, it will crash the worker process, significantly impacting customer service.
This is an optional functionality that should be safe to break. Worst case, it won't update concurrency.

Reply:
good catch